ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,356 @@
+use futures::stream::{iter as stream_iter, TryStreamExt};
+use ipa_macros::Step;
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, CustomArray, Field},
+    protocol::{
+        basics::Reveal, context::Context,
+        ipa_prf::boolean_ops::comparison_and_subtraction_sequential::compare_gt, RecordId,
+    },
+    secret_sharing::{replicated::semi_honest::AdditiveShare, WeakSharedValue},
+    seq_join::seq_join,
+};
+
+#[derive(Step)]
+pub(crate) enum Step {
+    Left,
+    Right,
+    Compare,
+    Reveal,
+}
+
+/// insecure quicksort using MPC comparisons and a key extraction function f
+///
+/// f takes as input an element in the slice and outputs the key by which we sort by
+/// follows partially the function signature of `sort_by_key`, see `https://doc.rust-lang.org/src/alloc/slice.rs.html#305-308`
+///
+/// set desc = true for descending ordering
+///
+/// This version of quicksort is insecure because it does not enforce the uniqueness of the sorted elements
+/// To see why this leaks information: assume that all elements are equal,
+/// then quicksort runs in time O(n^2) while for unique elements, it is only expected to run in time O(n log n)
+///
+/// The leakage can be fixed by appending a counter on each element that is unique to the element
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + Sync,
+    F: Fn(&S) -> &AdditiveShare<K> + Send + Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
+
+    // initialize stack
+    stack.push((ctx, 0usize, list.len()));
+
+    // iterate through quicksort recursions
+    while let Some((ctx, b_l, b_r)) = stack.pop() {
+        // start of quicksort function
+        // check whether sort is needed
+        if b_l + 1 < b_r {
+            // set up iterator
+            let mut iterator = list[b_l..b_r].iter();
+            // first element is pivot, apply key extraction function f
+            let pivot = f(iterator.next().unwrap());
+            // create pointer to context for moving into closure
+            let pctx = &ctx;
+            let pf = &f;
+            // precompute comparison against pivot and reveal result in parallel
+            let comp = seq_join(
```

Comment:
I don't understand why you're using seq_join here instead of a parallel join. As far as I can tell, there is no dependency between these computations, so they should all be able to be done in parallel.

Reply:
To my understanding seq_join is preferred when operating in parallel over multiple records of the same step while parallel join is preferred when operating in parallel over multiple steps of the same record. In this example, I need to finish a specific step (recursion) for all events (records) before I can move on to the next step (recursion). Alex suggested using try_join, I will sync with him.